keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,461 @@
+
+In this KEP, we are proposing a way to handle node shutdown cases that are not detected by Kubelet's Node Shutdown Manager. The pods will be forcefully deleted in this case, trigger the deletion of the VolumeAttachments, and new pods will be created on a new running node so that application can continue to function.
+
+Similarly, this approach can also be applied to the case when a node is in a non-recoverable state such as the hardware failure or broken OS.
+
+### Use Cases
+
+* User wants a stateful workload to failover to another running node in the following cases with minimum intervention when it's safe to do so.
+  * The node is shutdown
+  * The node is in other non-recoverable cases:
+    * hardware failure
+    * broken OS
+
+### Goals
+
+* Increase the availability of stateful workloads when a node is shutdown, or in other non-recoverable cases such as the hardware failure or broken OS.
+* Automate self-healing for stateful workloads.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.
+
+Existing logic:
+1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] This proposal requires a user to apply a `out-of-service` taint on a node when the user has confirmed that this node is shutdown or in a non-recoverable state due to the hardware failure or broken OS. Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used.
```

Comment:
I'd add something like: "The user setting the taint may use either the NoSchedule or the NoExecute effect, but since this proposal only affects terminating pods, they probably want NoExecute"

Reply:
If we add this to the KEP, does that mean in Pod GC Controller, we need to check the new taint which could be either NoSchedule and NoExecute effect?  Should we keep it simple and just use NoExecute?

Reply:
If a user set NoSchedule, then we don't do anything.

Reply:
I'm not getting this. If we don't do anything, then do we have to ask user to apply a new taint?

Reply:
When node is not ready, the taint with "NoExecute" effect will be added. So we cannot just check the taint effect, need to detect by its key "out-of-service" suggested here.

Reply:
node lifecycle controller automatically add the following node when kubelet is not responding 
node.kubernetes.io/not-ready: NoSchedule (need to confirm)
node.kubernetes.io/unreachable: NoExecute  --> normal eviction (gracefully delete)

Here we propose user to add new taint
node.kubernetes.io/out-of-service: NoExecute --> force delete

But the key name seems not easy to differentiate with unreachable, I feel.

Reply:
‘unreachable’ node will be back to the service if it is reachable again, but ‘out-of-service’ node won’t (or shouldn’t) because a user confirmed that it was down or had a problem and decided making it out-of-service.

Reply:
The point I am making is this:

There are several loosely-related things that need to happen for this KEP to tak eeffect:

1) stop scheduling to this node - any taint-effect will suffiice

2) evict some or all pods from this node - NoExecute will work, but the user might want a custom controller and why would we disallow that?

3) mark the node as out-of-service - using a taint for this is convenient since it works semantically

So why do we care which effect the out-of-service taint has?  All we care about is a) the node is OOS; b) a pod-with-volume is terminating.  NoExecute achieves both, but ANY mechanism that satisfies (a) and (b) should be sufficient, right?  Am I missing somethign?

Reply:
So if user add a taint "node.kubernetes.io/out-of-service:NoSchedule", what behavior should be expected? By definition, NoSchedule means
```
        // Do not allow new pods to schedule onto the node unless they tolerate the taint,
	// but allow all pods submitted to Kubelet without going through the scheduler
	// to start, and allow all already-running pods to continue running.
	// Enforced by the scheduler.
```

In this case, should controller still go ahead to force delete pods, then the behavior will be different from what is "NoSchedule" is defined?

Or maybe you are proposing the new logic in controller here should not directly trigger deleting pod first. If the OOS taint is added, it will check which pods are in terminating state, and force delete those pods? If somehow no pods are in terminating state, controller will not do anything?

Reply:
> Or maybe you are proposing the new logic in controller here should not directly trigger deleting pod first.

Exactly.  The magic of this KEP is that we get new information - "out-of-service", which we can intersect with "terminating" to do smarter / faster cleanup.

> If somehow no pods are in terminating state, controller will not do anything?

That sounds correct to me - maybe the user has some other process which will delete those pods in 30 seconds, so they set NoSchedule and let their own controller do it's thing.  That will eventually delete the pods, and the new (this KEP) logic kicks in.

Does that make sense?

Reply:
Got it. I think the above logic does provide more control knobs and flexibility. We need some document/guide for users to understand when they are enable this feature, what are the behaviors to expect in different situations.

Reply:
Is there any guidance for when a user should apply this taint?  I am trying to think about what automated system would apply this taint.

Reply:
In the "Proposed logic change" section, we mentioned that "Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used. "  An automated system can probably check how long a node has been NotReady.

Reply:
IMO,
* Before a user applies `out-of-service` taint to a shutdown node or a failed node, a user should confirm that it is powered off. If not, a user should make it powered off.
* Before a user turns on the power of a node with `out-of-service` taint, a user should confirm that there is no Pod and no attached PV on it with kubectl or so, even a user won’t make it returned a cluster.
* Before a user removes `out-of-service` taint from a node to restart hosting Pods, a user should confirm that:
    * its status is Ready with kubectl or so
    * there is no attached extra storage for PV on it with SSH or so